### PR TITLE
fix: migrate configs when file path changes

### PR DESCRIPTION
**Pull request by Buckwich**
Closes https://github.com/camunda/camunda-modeler/issues/5360

### Proposed Changes

- align reading configs to the same set behavior for unsaved files
- migrate configs if the path changes (eg. unsaved `null` -> `new.bpmn` or saveAs `old.bpmn` -> `new.bpmn`

Once task testing is also no longer using null as a config key they can be added to the migration keys

Solves issues were connections were not loaded for unsaved or "renamed" (saveAs) files

https://github.com/user-attachments/assets/0410505a-c89e-41ea-bf8e-61bf810c93f7

### Alternatives considered

- Problem: connection information is linked to the filepath, all unsaved file share that information (previously key was null, currently separate config key)
- First approach: Connection manager listens on tab saved and rewrites connection information. Works for deploy and start instance but not task testing due to asynchrosity
- Second: let the deployment tool handle the save logic and ensure connection is set. works in most cases but not if the user changes the file path (save as)

### Checklist

Ensure you provided everything we need to review your contribution:

* [x] Your __contribution meets the [definition of done](https://github.com/bpmn-io/.github/blob/main/resources/DEFINITION_OF_DONE.md)__
* [x] Any new additions or modifications are __consistent with the existing UI and UX patterns__
* [ ] __Pull request description establishes context__:
  * [x] __Link to related issue(s)__, i.e. `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`
  * [x] __Brief textual description__ of the changes
  * [x] __Screenshots or short videos__ showing UI/UX changes
  * [ ] __Steps to try out__, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)

### CODE REVIEW

File: client/src/app/App.js [hunk 1]
```diff
@@ -1445,14 +1445,43 @@ export class App extends PureComponent {
...
+      const value = await (oldPath
+        ? config.getForFile({ path: oldPath }, key)
+        : config.get(key));
+
+      if (value) {
+        await config.setForFile({ path: newPath }, key, value);
+      }
+
+    }
   }
```
[the remark this answers is not in the document]
Comment (Buckwich) [reply client/src/app/App.js:1486]: sorry forgot to commit them

File: client/src/app/App.js [hunk 2]
```diff
@@ -1445,14 +1445,43 @@ export class App extends PureComponent {
...
+  /**
+   * Migrate configurations from old file path to new file path.
+   *
+   * @param {string} oldPath - Original file path
+   * @param {string} newPath - New file path
+   */
+  async migrateConfigForFilePath(oldPath, newPath) {
+    const config = this.getGlobal('config');
+
+    const keysToMigrate = [ 'connection-manager' ];
```
[the remark this answers is not in the document]
Comment (Buckwich) [reply client/src/app/App.js]: done

File: client/src/app/App.js [hunk 3]
```diff
@@ -1445,14 +1445,43 @@ export class App extends PureComponent {
...
+      await this.migrateConfigForFilePath(originalFile.path, savePath);
+    }
+
+    return file;
+  }
+
+  /**
+   * Migrate configurations from old file path to new file path.
+   *
+   * @param {string} oldPath - Original file path
```
[the remark this answers is not in the document]
Comment (Buckwich) [reply client/src/app/App.js]: ok

File: client/src/app/App.js [hunk 4]
```diff
@@ -1445,14 +1446,43 @@ export class App extends PureComponent {
...
+  /**
+   * Migrate configurations from old file path to new file path.
+   *
+   * @param {string|null} oldPath - Original file path (null for new/unsaved files)
+   * @param {string} newPath - New file path
+   */
+  async migrateConfigForFilePath(oldPath, newPath) {
+    const config = this.getGlobal('config');
+
+    const keysToMigrate = [ CONFIG_KEYS.CONNECTION_MANAGER ];
```
Comment (philippfromme) [inline client/src/app/App.js:1473]: Is there a reason to only migrate specific configs? Shouldn't we migrate all configs scoped to that file?

For example you might have this for an unsaved file:

```
"null": {
  "taskTesting": {
    "input": {
      "Task_1": "{\n  \"foo\": \"bar\"\n}"
    },
    "output": {}
  }
},
```

Comment (Buckwich) [reply client/src/app/App.js:1473]: i was not sure what kind of side effects i would get with task testing as its using null to store the config for unsaved files. connection manager uses a key outside of files for unsaved files

Comment (Buckwich) [reply client/src/app/App.js:1473]: I tried to make it generic, but without the key array I don't know if I need to migrate something.

eg I could have this config:
```javascript
{
  connection-manager: {
    id: "defaultForUnsaved"
  }
  files:{}
}
```

if i save a file I want it to be:
```
{
  connection-manager: {
    id: "defaultForUnsaved"
  }
  files:{
    newFile: {
     connection-manager: {
       id: "defaultForUnsaved"
     }
  }
}
```

Comment (Buckwich) [reply client/src/app/App.js:1473]: @barmac what do you think about moving the top level key to something like `fileDefaults`? Then we wouldn't need the key array
```
{
  fileDefaults: {
    connection-manager: {
      id: "defaultForUnsaved"
    }
  }
  files:{}
}
```

Comment (Buckwich) [reply client/src/app/App.js:1473]: alternative could be https://github.com/camunda/camunda-modeler/pull/5464/files#diff-b650ad882ccc34a87aa822cb69b9bffc2e3c1513db7dbdff0fc8aaec11fd99a2

Comment (philippfromme) [reply client/src/app/App.js:1473]: I wasn't aware a default is saved outside of `files`. My assumption was that it would like something like this for unsaved files:

```
{
  "files": {
    "null": {
      "connection-manager": { ... },
      "task-testing": { ... }
    }
  }
}
```

So that would be the de facto default for unsaved files. On the other hand especially for task testing that wouldn't make sense. I wouldn't want to reuse the task testing configuration for two unsaved diagrams. They would also override each others configuration constantly. I guess there are two aspects:

1. Being able to save file-specific feature X configuration for unsaved files without conflicts
2. The default configuration for unsaved files as defined by feature X

Comment (Buckwich) [reply client/src/app/App.js:1473]: that was the original implementation, but @barmac didn't like writing to null path. We also discussed to adjust task testing to not rely on null (https://github.com/camunda/camunda-modeler/pull/5422#issuecomment-3548528951)

Comment (philippfromme) [reply client/src/app/App.js:1473]: Opened an issue for 1. https://github.com/camunda/camunda-modeler/issues/5468

Comment (philippfromme) [reply client/src/app/App.js:1473]: So my suggestion would be to not add a `migrateConfigForFilePath` with hard-coded configs but to solve 1. and simply apply a default when a new file is created. Migration would then not be necessary. Once the file is saved the config is saved with a new key which is now the path.

Comment (Buckwich) [reply client/src/app/App.js:1473]: i haven't tested it with task testing but for connection manager using save as would not persist the change. in the UI it would still look like it because its part of the application state but on reload it would be lost

Comment (Buckwich) [reply client/src/app/App.js:1473]: so we would still need something like https://github.com/camunda/camunda-modeler/pull/5464/files#diff-b650ad882ccc34a87aa822cb69b9bffc2e3c1513db7dbdff0fc8aaec11fd99a2

Comment (philippfromme) [reply client/src/app/App.js:1473]: This mechanism could then also be used by any other feature. When a new file is created, I want to add configuration X and once I've done that I never need to worry about what happens once the file is saved.

Comment (philippfromme) [reply client/src/app/App.js:1473]: > so we would still need something like https://github.com/camunda/camunda-modeler/pull/5464/files#diff-b650ad882ccc34a87aa822cb69b9bffc2e3c1513db7dbdff0fc8aaec11fd99a2

Kind of like that, but the defaults would by applied to any new file immediately.

Comment (barmac) [reply client/src/app/App.js:1473]: > I wouldn't want to reuse the task testing configuration for two unsaved diagrams. 

💯 This is why we _don't_ want to use `null` file config.

Comment (philippfromme) [reply client/src/app/App.js:1473]: I'm wondering why we even write to the `config.json` for unsaved files since they will be gone if you don't save them before closing the modeler. Could look something like this:

```
// In-memory storage for unsaved files
const unsavedFileSettings = new Map();

async getForFile(file, key, defaultValue = null) {
  const { path } = file;

  if (!path) {
    const configForFile = unsavedFileSettings.get(file.id) || {};
    return key ? configForFile[key] : configForFile;
  }

  const files = await this.get('files') || {};
  const configForFile = files[path];

  if (!configForFile) {
    return defaultValue;
  }

  return key ? configForFile[key] : configForFile;
}

async setForFile(file, key, value) {
  const { path } = file;

  if (!path) {
    const configForFile = unsavedFileSettings.get(file.id) || {};
    configForFile[key] = value;
    unsavedFileSettings.set(file.id, configForFile);
    return configForFile;
  }

  const files = await this.get('files') || {};
  const configForFile = files[path] = files[path] || {};
  configForFile[key] = value;
  await this.set('files', files);
  return configForFile;
}
```

Comment (Buckwich) [reply client/src/app/App.js:1473]: the idea in this pr is that the user can set the default (persisted) the same way they would set any other connection. but its not really obvious that its changing it for all unsaved files

Comment (Buckwich) [reply client/src/app/App.js:1473]: for users to be more explicit that could be another setting: 
image

Comment (philippfromme) [reply client/src/app/App.js:1473]: > they would set any other connection. but its not really obvious that its changing it for all unsaved files

So it's more like last used.
